core/types/block.go

```diff
@@ -527,6 +555,27 @@ func (h *Header) SetExcessDataGas(v *big.Int) {
 // Hash returns the block hash of the header, which is simply the keccak256 hash of its
 // RLP encoding.
 func (h *Header) Hash() libcommon.Hash {
+	// The difficult of legacy blocks is 1 or 2
+	if h.Difficulty.Cmp(libcommon.Big1) == 0 || h.Difficulty.Cmp(libcommon.Big2) == 0 {
```

Comment:
Probably fine... but should we not check the chainID or something instead?

Reply:
There is no field like chain ID in header. The difficulty might be the easy thing for us to distinguish the legacy block and new block. The legacy block always have `difficulty = 1 or 2`. The new block is mined by POS, so the difficulty is always 0.